### ISSUE TITLE
Differences in results with pure R and `umap-learn` calls

### ISSUE THREAD

**Issue by adomingues** [tkonopka/umap#26]
Hi, I am finding it hard to create the same embedding with the pure R and the python API call using `method = "umap-learn"`. Using the `iris` dataset, the differences are fairly small, but noticeable, and when using a larger dataset it becomes much larger. I am hopping that you clarify where differences come from and which settings I am missing.

# Iris dataset

Note that I am using the default arguments in both calls assuming that the defaults are the same.

```r
# Source specific lines in an R file
# https://gist.github.com/christophergandrud/1eb4e095974204b12af9
# @param file character string with the path to the file to source.
# @param lines numeric vector of lines to source in \code{file}.

source_lines <- function(file, lines){
    source(textConnection(readLines(file)[lines]))
}

source_lines(
  "https://raw.githubusercontent.com/cran/umap/refs/heads/master/inst/doc/umap.R",
  c(5:38)
)

library("umap")
library("data.table")
library("dplyr")
library("ggplot2")
library("magrittr")

iris.data <- iris[, grep("Sepal|Petal", colnames(iris))]
iris.labels <- iris[, "Species"]

```

First the R only implementation:

```r
iris.umap <- umap(iris.data, random_state = 42)
plot.iris(iris.umap, iris.labels)

```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/03596bb0-1325-4b35-a040-afb012b307ac" />

And changing to the `umap-learn` backend (0.5.9.post2):

```r
iris.umap <- umap(iris.data, method = "umap-learn", random_state = 42)
plot.iris(iris.umap, iris.labels)

```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/01b8cd39-a1f5-4c7b-84f0-13175c4e02df" />

Not such a massive difference, though I would expect that by setting `random_state = 42` I would ensure reproducibility between the two implementations.

# L1000 dataset

With this larger data set things get more interesting in that the differences are somewhat large.

```r
# data prep
l1000 <- fread("https://github.com/broadinstitute/lincs-profiling-complementarity/raw/5091585bde14c33d6819fa6b494e352a064fbeb4/1.Data-exploration/Profiles_level4/L1000/L1000_lvl4_cpd_replicate_datasets/L1000_level4_cpd_replicates.csv.gz") %>%
  as.data.frame()

l1000_probes <- l1000 %>%
  select(ends_with("_at"))
l1000_features <- l1000 %>%
  select(-ends_with("_at"))


pc <- prcomp(
  l1000_probes,
  center = TRUE,
  scale. = TRUE
)
pca_mat <- pc$x[, 1:50]

```

Pure `R`:

```r
l100_umap <- umap(pca_mat)

ummap_r_df <- cbind(
  l100_umap$layout,
  l1000_features
) %>%
  rename(
    "UMAP_1" = 1,
    "UMAP_2" = 2
  )

ummap_r_df %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = moa)) +
  geom_point() +
  theme(legend.position = "none")

```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/f9cdf285-8623-4b60-87df-e9c4b8866bd8" />

And changing to the `umap-learn` backend (0.5.9.post2):

```r
l100_umap <- umap(pca_mat, method = "umap-learn", random_state = 42)

ummap_r_df <- cbind(
  l100_umap$layout,
  l1000_features
) %>%
  rename(
    "UMAP_1" = 1,
    "UMAP_2" = 2
  )

ummap_r_df %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = moa)) +
  geom_point() +
  theme(legend.position = "none")

```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/50c83883-756e-4be6-a6b6-a89f81c7f82e" />

The structure is sort of ok, but the embedding range is different and the both the global and local structure have notable differences.

If you could point me out to which settings are different in the implementations and cause these differences it would be appreciated.

Cheers.

**Comment by tkonopka** [comment]
Thanks @adomingues for the detailed analysis.

In short, one cannot expect exact equality between the R and python implementations. And more generally, it is important to mention that variability can appear between runs of one dimensional reduction approach, so outputs must always be interpreted with care.

In more detail:

- The R and python implementations are written and maintained separately, and they have been around for several years now. The approaches had slight differences at the beginning, and there may be others that appeared since then.
- Both R and python implementations rely on stochasticity to run efficiently. The R and python languages use different default generators, so their behaviour and outputs will diverge slightly just because of that. This is normal - indeed similar variability can be observed by running one algorithm with different seeds.

Either point above by itself means that it is not possible to achieve exact same outputs from the R and python implementations (there are no special settings). Nevertheless, the two implementations can still be called "umap" with full honesty (imo). Underneath, the algorithms have many common properties, with the R version being a translation of early python code. If you like, you can think of them as two members of a "family" of dimensional reduction methods.

As mentioned in the second point, variations appear even when using a single algorithm and changing the seed for the random number generator, so dealing with the variability is part of using dimensional reduction. The variability can sometimes change interpretation of a dataset, which is unfortunate but it is a weakness of dimensional reduction in general. I would recommend to always trial a few embeddings to avoid reading too much into a single embedding. And if you are looking to publish an analysis so that others can reproduce it, it is best to share as much detail about the computational environment as possible, including package versions and exact code, because the name of the algorithm alone is not sufficient for exact matching.

**Comment by adomingues** [comment]
Hi @tkonopka, 

thanks a lot for the detailed reply. I am aware of most of the issues and subtleties you pointed out, but I was hopeful the differences would be less pronounced when using the same parameters :) 

Also, and to be clear, I am not saying that is a “correct” UMAP implementation, every author will pick different defaults (with good reasons). My goal was more trying to unpick where the (meaningful) differences come from.

Importantly, I held out the major difference I have encountered because I could not replicate it properly - now I can and its quite something.  I have observed it a couple of years ago in a different system, but never got around to dig deeper, which I finally did. Since I am not sure this is even a `umap` issue, please feel free to say say and we close this issue. Full description below.

---

Using the same large dataset from before I use python to read the data and do the PCA. The key aspect is that I am using `reticulate` to call python which might complicate things as we’ll see below:

```R
Sys.setenv(RETICULATE_PYTHON_ENV = "/home/adomingues/Documents/projects/datalyst/posts/2025-11-25-ump-r-python/.venv") # I can't have nice things
library("reticulate")
library("umap")
library("dplyr")
library("ggplot2")
library("magrittr")
set_theme(theme_light())

reticulate::py_config()

```

> python:         /home/adomingues/Documents/projects/datalyst/posts/2025-11-25-ump-r-python/.venv/bin/python

> libpython:      /usr/lib/python3.12/config-3.12-x86_64-linux-gnu/libpython3.12.so
> pythonhome:     /home/adomingues/Documents/projects/2025-11-25-ump-r-python/.venv:/home/adomingues/Documents/projects/2025-11-25-ump-r-python/.venv
> virtualenv:     /home/adomingues/Documents/projects/2025-11-25-ump-r-python/.venv/bin/activate_this.py
> version:        3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]
> numpy:          /home/adomingues/Documents/projects/2025-11-25-ump-r-python/.venv/lib/python3.12/site-packages/numpy
> numpy_version:  2.3.5
> umap:           /home/adomingues/Documents/projects/2025-11-25-ump-r-python/.venv/lib/python3.12/site-packages/umap
> 
> NOTE: Python version was forced by RETICULATE_PYTHON_ENV

Data import and PCA dimensionality reduction is done in `python`:

```python
import pathlib
import numpy as np
import pandas as pd
import umap
from sklearn.decomposition import PCA
import plotnine as pl

# Load L1000 profiles
file = "https://github.com/broadinstitute/lincs-profiling-complementarity/raw/5091585bde14c33d6819fa6b494e352a064fbeb4/1.Data-exploration/Profiles_level4/L1000/L1000_lvl4_cpd_replicate_datasets/L1000_level4_cpd_replicates.csv.gz"
df = pd.read_csv(file, low_memory=False)

features = df.columns[df.columns.str.endswith("at")].tolist()
meta_features = df.drop(features, axis="columns").columns.tolist()
features_df = df.loc[:, meta_features]
data_df = df.loc[:, features]
# Transform PCA to top 200 components
n_components = 200
pca = PCA(n_components=n_components)

pca_df = pca.fit_transform(data_df)
pca_df = pd.DataFrame(pca_df)
pca_df.columns = [f"PCA_{x}" for x in range(0, n_components)]

```

The top 200 principal components are then passed to `R` for `umap` reduction and plotting:

```R
pca_py <- py$pca_df

umap_r <- umap::umap(
  pca_py,
  random_state = 123,
  min_dist = 0.1,
  spread = 1,
  n_neighbors = 25,
  metric = "euclidean"
)

umap_r_p <- cbind(
  umap_r$layout,
  py$features_df
) %>%
  rename(
    "UMAP_1" = 1,
    "UMAP_2" = 2
  ) %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = moa)) +
  geom_point() +
  labs(
    title = "PCA in python, umap::umap"
    ) +
  theme(legend.position = "none")

umap_r_p
```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/67bc7d23-be35-4344-bf63-146117ac5227" />

Now this is wild. The range of the embedding alone is quite large. But when I run the same data with method = "umap-learn" and the result is what I expected (and replicates the paper I am reproducing):

```R
umap_learn_r <- umap::umap(
  pca_py,
  random_state = 123,
  min_dist = 0.1,
  spread = 1,
  n_neighbors = 25,
  metric = "euclidean",
  method = "umap-learn"
)

umap_learn_r_p <- cbind(
  umap_learn_r$layout,
  py$features_df
) %>%
  rename(
    "UMAP_1" = 1,
    "UMAP_2" = 2
  ) %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = moa)) +
  geom_point() +
  labs(
    title = "PCA in python, UMAP-learn in R"
    ) +
  theme(legend.position = "none")
umap_learn_r_p
```

<img width="1344" height="960" alt="Image" src="https://github.com/user-attachments/assets/4e6a9e9e-a845-4c60-8127-000a1b5c264f" />

And this is what I cannot explain. It is possible that `reticulate` is somehow messing my environment, but then it should affect the `pyton`  `umap-learn` not the pure R implementation. Note that in my previous message I used R without calling python via reticulate, or at all.

I totally knowledge this might not be something related to `umap` at all, but since it's so odd you might want to know about it. And it's entirely possible that I made some obvious mistake, but I can't spot it at all.

**Comment by tkonopka** [comment]
Yes, those are quite different. It is possibly related to the clumpiness of the points in the dataset that make the initialization of the layout very spread out, before the umap algorithm kicks in to arrange the points. One way to make the effect less pronounced could be to initialize the layout manually, e.g. passing the first two pca components as an initial layout (argument `init=as.matrix(pcapy[, 1:2])`). Another way would be to make the input dataset less clumpy, e.g. using fewer PCA components to begin with. Finally, sometimes it can be interesting to trim out outliers and zoom in on a subset of the data, where the layout will likely look less pathological. I realize all these suggestions are imperfect in their own way, but I probably won't have time to adjust the package itself to alleviate this behaviour.

**Comment by adomingues** [comment]
Thank you for getting back to me @tkonopka - I appreciate the time you took to look into this. I will keep those options in mind when I use `umap` next time.
